File: backend/app/schemas/analytics.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class EventFeedItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    recipient: str
    event_type: str
    timestamp: datetime
    ip_address: str | None = None
    user_agent: str | None = None


class AnalyticsResponse(BaseModel):
    total_emails: int
    open_rate: float
    click_rate: float
    events: list[EventFeedItem]

File: frontend/src/lib/analytics.ts
import { apiFetch } from "@/lib/api";

export interface EventFeedItem {
  recipient: string;
  event_type: "open" | "click" | string;
  timestamp: string;
  ip_address: string | null;
  user_agent: string | null;
}

export interface AnalyticsResponse {
  total_emails: number;
  open_rate: number;
  click_rate: number;
  events: EventFeedItem[];
}

export async function fetchAnalytics(): Promise<AnalyticsResponse> {
  return apiFetch<AnalyticsResponse>("/api/analytics");
}
